### PR TITLE
Fix contract interaction recipient and contract address

### LINKED ISSUE #18800: [Bug]: Contract Interaction pages should show the contract address as the recipient
### Describe the bug

Reported by @seaona in https://github.com/MetaMask/metamask-extension/pull/18652#issuecomment-1519635122

For contract interaction pages, the recipient should be the contract address. 

In this example, the `approve` method is being called with a sent ETH value, which results in the fallback contract interaction page rather than the custom approve page shown. Here, "Address 2" should be replaced by the contract address.
image

Maybe we can consider showing "Contract Interaction" instead of these resolved names in the summary header as well

### Steps to reproduce

Follow the steps to repro with Remix here: https://github.com/MetaMask/MetaMask-planning/issues/380

or

use Sepholia contract with example payable token methods here:
https://sepolia.etherscan.io/address/0x07ffca1de83676dca804126f45a0d14b3e037fad

### Error messages or log output

_No response_

### Version

10.29

### Build type

None

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by digiwand**
## Explanation

Fixes #18800

A contract interaction transaction may send value along with it. When this happens, we display the contract interaction screen instead of the custom transaction screens. This is the case with types: approve, setApprovalForAll, revoke, transfer, and safeTransferFrom. We should display the contract address as the recipient in these cases. This PR fixes this. 

Other updates: 
- add HexStringUtil#isEmptyHexString
- adds tests
- add/update TransactionDetail row item keys

Note, there is an occurrence where the transaction may be treated as a simpleSend type, and "New Contract" may show instead of the deployed contract address. This will be addressed with #18801

---

Adding tests has caught this console:error :
> Warning: Each child in a list should have a unique "key" prop
Screen Shot 2023-05-02 at 2 39 11 PM

which prompted the "add/update TransactionDetail row item keys" update (https://github.com/MetaMask/metamask-extension/pull/18855/commits/a20e957e7d776d921ae665f97f3f5d5f1166204d).

### Before

Screen Shot 2023-04-27 at 1 53 54 AM
Screen Shot 2023-04-27 at 1 54 42 AM
Screen Shot 2023-04-27 at 1 55 56 AM

### After

Screen Shot 2023-04-27 at 12 53 58 PM
Screen Shot 2023-04-27 at 10 29 00 AM
Screen Shot 2023-04-27 at 10 09 43 AM

## Manual Testing Steps

Follow the steps to repro with Remix here: https://github.com/MetaMask/MetaMask-planning/issues/380 and observe the recipient address

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Review by jpuri — DISMISSED** [review]
Great work 👍

**Review by DDDDDanica — DISMISSED** [review]

**Review by DDDDDanica — DISMISSED** [review]

**Review by jpuri — DISMISSED** [review]
🚀

File: ui/pages/confirm-transaction-base/confirm-transaction-base.test.js [hunk 1]
```diff
@@ -155,15 +181,92 @@ describe('Confirm Transaction Base', () => {
...
+        const store = configureMockStore(middleware)(mockedStore);
+        const { container } = renderWithProvider(
+          <ConfirmTransactionBase actionKey="confirm" />,
+          store,
+        );
+
+        const recipientElem = container.querySelector(sendToRecipientSelector);
+        expect(recipientElem).toHaveTextContent(mockTxParamsToAddressConcat);
+      });
+    });
```
Comment (jpuri) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.test.js:270]: 👍 nice unit test coverage

**Comment by digiwand** [comment]
Hey @seaona (and other QA'ers), this is ready to be QA'd. 

The issue https://github.com/MetaMask/metamask-extension/issues/18801 still exists, but seems to be hard to reproduce. If you are able to uncover more details on repro'ing this issue where a contract interaction with a value may resolve as a send transaction rather than a contract interaction, please share notes. Thank you! 🙏

**Comment by digiwand** [comment]
Thanks for the catch @seaona! Fixed the issue and updated the tests.

**Comment by seaona** [comment]
I see the issue fixed! Thank you @digiwand ! PR is ready to be merged :100:

**Review by jpuri — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

**Review by DDDDDanica — APPROVED** [review]

**Pull request merged**